File: Leetcode_Python/meetingsRoomsii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        startTimes = sorted([i[0] for i in intervals])
        endTimes = sorted([i[1] for i in intervals])

        start, end = 0, 0

        noOfRooms = 0
        noOfMeetings = 0

        while start < len(intervals):
            if startTimes[start] < endTimes[end]:
                start += 1
                noOfMeetings += 1
            else:
                end += 1
                noOfMeetings -= 1

            noOfRooms = max(noOfRooms, noOfMeetings)

        return noOfRooms

File: Leetcode_Java/meetingsRoomsii.java
import java.util.*;

class Solution {
    public int minMeetingRooms(int[][] intervals) {
        int n = intervals.length;
        int[] startTimes = new int[n];
        int[] endTimes = new int[n];

        for (int i = 0; i < n; i++) {
            startTimes[i] = intervals[i][0];
            endTimes[i] = intervals[i][1];
        }

        Arrays.sort(startTimes);
        Arrays.sort(endTimes);

        int start = 0, end = 0;
        int noOfRooms = 0, noOfMeetings = 0;

        while (start < n) {
            if (startTimes[start] < endTimes[end]) {
                start++;
                noOfMeetings++;
            } else {
                end++;
                noOfMeetings--;
            }

            noOfRooms = Math.max(noOfRooms, noOfMeetings);
        }

        return noOfRooms;
    }
}
